pkg/operatorstatus/builder.go

```diff
@@ -0,0 +1,190 @@
+	return b
+}
+
+// WithDegraded sets an OperatorDegraded type condition.
+func (b *Builder) WithDegraded(status configv1.ConditionStatus, reason string) *Builder {
+	b.init()
+	condition := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorDegraded,
+		Status:             status,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+		Reason:             reason,
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithAvailable sets an OperatorAvailable type condition.
+func (b *Builder) WithAvailable(status configv1.ConditionStatus, message string) *Builder {
+	b.init()
+	condition := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorAvailable,
+		Status:             status,
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithVersion adds the specific version into the status.
+func (b *Builder) WithVersion(name, version string) *Builder {
+	b.init()
+
+	existing := b.findVersion(name)
+	if existing != nil {
+		existing.Version = version
+		return b
+	}
+
+	ov := configv1.OperandVersion{
+		Name:    name,
+		Version: version,
+	}
+	b.status.Versions = append(b.status.Versions, ov)
+
+	return b
+}
+
+// WithRelatedObject adds the reference specified to the RelatedObjects list.
+func (b *Builder) WithRelatedObject(group, resource, namespace, name string) *Builder {
+	b.init()
+
+	reference := configv1.ObjectReference{
+		Group:     group,
+		Resource:  resource,
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	b.setRelatedObject(reference)
+
+	return b
+}
+
+func (b *Builder) init() {
+	if b.status == nil {
+		b.status = &configv1.ClusterOperatorStatus{
+			Conditions:     []configv1.ClusterOperatorStatusCondition{},
+			Versions:       []configv1.OperandVersion{},
+			RelatedObjects: []configv1.ObjectReference{},
+		}
+	}
+}
+
+func (b *Builder) findCondition(conditionType configv1.ClusterStatusConditionType) *configv1.ClusterOperatorStatusCondition {
+	for i := range b.status.Conditions {
+		if b.status.Conditions[i].Type == conditionType {
+			return &b.status.Conditions[i]
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) setCondition(condition *configv1.ClusterOperatorStatusCondition) {
+	existing := b.findCondition(condition.Type)
+	if existing == nil {
+		b.status.Conditions = append(b.status.Conditions, *condition)
+		return
+	}
+
+	existing.Reason = condition.Reason
+	existing.Message = condition.Message
+
+	if existing.Status != condition.Status {
+		existing.Status = condition.Status
+		existing.LastTransitionTime = condition.LastTransitionTime
+	}
+}
+
+func (b *Builder) findVersion(name string) *configv1.OperandVersion {
+	for i := range b.status.Versions {
+		if b.status.Versions[i].Name == name {
+			return &b.status.Versions[i]
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) setRelatedObject(reference configv1.ObjectReference) {
+	for i := range b.status.RelatedObjects {
```

Comment:
Is this really required? Without this check this function would be `setRelatedObject`